results/cmd/watcher/main_test.go

```diff
@@ -0,0 +1,112 @@
+)
+
+func getRunName(tr *v1beta1.TaskRun) string {
+	return strings.Join([]string{tr.Namespace, tr.Name}, "/")
+}
+
+// getFakeClients create a fake client to send test data to reconciler
+func getFakeClients(t *testing.T, tr []*v1beta1.TaskRun, conn *grpc.ClientConn) test.Assets {
+	t.Helper()
+	ctx, _ := ttesting.SetupFakeContext(t)
+	d := test.Data{
+		TaskRuns: tr,
+	}
+	clients, _ := test.SeedTestData(t, ctx, d)
+	cmw := configmap.NewInformedWatcher(clients.Kube, "")
+	return test.Assets{
+		Controller: newController(ctx, cmw, conn),
+		Clients:    clients,
+	}
+}
+
+// reconcileTaskRun sends TaskRun data to reconciler and then retrieves completed TaskRun
+func reconcileTaskRun(ctx context.Context, asset test.Assets, taskRun v1beta1.TaskRun) (*v1beta1.TaskRun, error) {
+	c := asset.Controller
+	clients := asset.Clients
+	if err := c.Reconciler.Reconcile(ctx, getRunName(&taskRun)); err != nil {
+		return nil, err
+	}
+	tr, err := clients.Pipeline.TektonV1beta1().TaskRuns(taskRun.Namespace).Get(taskRun.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return tr, nil
+}
+
+func setupServerConnection(t *testing.T) (*grpc.ClientConn, error) {
+	srv, err := server.SetupTestDB(t)
+	if err != nil {
+		t.Fatalf("Failed to create fake server: %v", err)
+	}
+	s := grpc.NewServer()
+	pb.RegisterResultsServer(s, srv) // local test server
+	lis, err := net.Listen("tcp", port)
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+	go s.Serve(lis)
+	t.Cleanup(func() {
+		lis.Close()
+		s.Stop()
+	})
+	return grpc.Dial(*apiAddr, grpc.WithInsecure(), grpc.WithBlock())
+}
+
+// TestReconcile tests if TaskRun in the client can be properly updated when sent to reconciler
+func TestReconcile(t *testing.T) {
+	conn, err := setupServerConnection(t)
+	if err != nil {
+		t.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+	taskRun := v1beta1.TaskRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "Tekton",
+			Namespace:   "default",
+			Annotations: map[string]string{"demo": "demo"},
+		},
+	}
+
+	asset := getFakeClients(t, []*v1beta1.TaskRun{&taskRun}, conn)
+	ctx := context.Background()
+	
+	// Create a new TaskRun in the fake client, completed TaskRun should be updated with a results id field in annotations.
+	tr, err := reconcileTaskRun(ctx, asset, taskRun)
+	if err != nil {
+		t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+	}
+	if _, ok := tr.Annotations[idName]; !ok {
+		t.Fatalf("Expected completed TaskRun %s should be updated with a results_id field in annotations", taskRun.Name)
+	}
+	// Update a TaskRun with results id, completed TaskRun should remain unchanged.
+	newtr, err := reconcileTaskRun(ctx, asset, *tr)
+	if err != nil {
+		t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+	}
+	if diff := cmp.Diff(tr, newtr); diff != "" {
+		t.Fatalf("Expected completed TaskRun should remain unchanged when it has a results_id in annotations: %v", diff)
+	}
+}
```

Comment:
This does a good job of testing the new build flow, but we should also add a test for the update flow as well.

Reply:
Added a update-flow test